### PR TITLE
Toniof 522 provide versions

### PR DESCRIPTION
**Pull request by TonioF**
Solves #522. I have written the function in a way that it can be re-used by Cate.

Please see whether the output is in all the places where you would expect it.

Checklist:

* [x] Add unit tests and/or doctests in docstrings
* [x] Add docstrings and API docs for any new/modified user-facing classes and functions
* ~[ ] New/modified features documented in `docs/source/*`~
* [x] Changes documented in `CHANGES.md`
* [x] AppVeyor CI passes
* [x] Test coverage remains or increases (target 100%)

Remember to close associated issues after merge!

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -83,6 +83,8 @@
 * Added a new utility module `xcube.util.temp` that allows for creating 
   temporary files and directories that will be deleted when the current 
   process ends.
+* Added functions `get_dependencies` and `get_xcube_dependencies` to get
```
Comment (forman) [inline CHANGES.md:86]: Where are they. Pls provide fully qualified names.

File: xcube/cli/gen2.py [hunk 2]
```diff
@@ -147,6 +148,7 @@ def gen2(request_path: str,
 
         result = dict(status='error',
                       message=f'{error}',
+                      dependencies=get_xcube_dependencies(),
```
Comment (forman) [inline xcube/cli/gen2.py:151]: Please use `versions`

File: xcube/core/gen2/local/generator.py [hunk 3]
```diff
@@ -226,15 +227,17 @@ def __generate_cube(self, request: CubeGeneratorRequest) \
...
             )
         else:
             return CubeGeneratorResult(
                 status='warning',
                 status_code=422,
                 message=f'An empty cube has been generated'
                         f' after {total_time:.2f} seconds.'
-                        f' No data has been written at all.'
+                        f' No data has been written at all.',
+                dependencies = get_xcube_dependencies()
```
Comment (forman) [inline xcube/core/gen2/local/generator.py:240]: Only include on error

File: xcube/core/gen2/local/generator.py [hunk 4]
```diff
@@ -226,15 +227,17 @@ def __generate_cube(self, request: CubeGeneratorRequest) \
                 status_code=201,
                 result=CubeReference(data_id=data_id),
                 message=f'Cube generated successfully'
-                        f' after {total_time:.2f} seconds'
+                        f' after {total_time:.2f} seconds',
+                dependencies=get_xcube_dependencies()
```
Comment (forman) [inline xcube/core/gen2/local/generator.py:231]: Only include on error

File: xcube/core/gen2/local/generator.py [hunk 5]
```diff
@@ -245,4 +248,5 @@ def _get_cube_info(self, request: CubeGeneratorRequestLike) \
         cube_info = informant.generate()
         return CubeInfoResult(result=cube_info,
                               status='ok',
-                              status_code=200)
+                              status_code=200,
+                              dependencies=get_xcube_dependencies())
```
Comment (forman) [inline xcube/core/gen2/local/generator.py:252]: Only include on error

File: xcube/core/gen2/response.py [hunk 6]
```diff
@@ -52,6 +53,7 @@ def __init__(self,
         self.message = message if message else None
         self.output = list(output) if output else None
         self.traceback = list(traceback) if traceback else None
+        self.dependencies = dependencies if dependencies else None
```
Comment (forman) [inline xcube/core/gen2/response.py:56]: use name `versions`

File: xcube/core/gen2/response.py [hunk 7]
```diff
@@ -87,6 +91,7 @@ def get_schema(cls) -> JsonObjectSchema:
                 message=JsonStringSchema(),
                 output=JsonArraySchema(items=JsonStringSchema()),
                 traceback=JsonArraySchema(items=JsonStringSchema()),
+                dependencies=JsonObjectSchema(additional_properties=True)
```
Comment (forman) [inline xcube/core/gen2/response.py:94]: use name `versions`

File: xcube/util/dependencies.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+DEPENDENCY_NAMES = \
+    ['affine', 'click', 'cmocean', 'dask', 'dask - image', 'distributed',
+     'fiona', 'fsspec', 'gdal', 'geopandas', 'jdcal', 'jsonschema',
+     'matplotlib - base', 'netcdf4', 'numba', 'numpy', 'pandas', 'pillow',
+     'pyjwt', 'pyproj', 'pyyaml', 'rasterio', 'requests',
+     'requests - oauthlib', 's3fs', 'scipy', 'setuptools', 'shapely',
+     'tornado', 'urllib3', 'xarray', 'zarr']
+
+
+def get_xcube_dependencies() -> Dict[str, str]:
```
Comment (forman) [inline xcube/util/dependencies.py:38]: * rename function to `get_xcube_versions()`
* rename module to `versions.py`

**Review by forman — CHANGES_REQUESTED** [review]
Please rename `dependencies` to `versions`. 

The original idea was to just copy the corresponding function from Cate.
Im not a fan of including any dependency in our responses. This would be another operatioon of the generator API  and the xcube server API.

**Comment by TonioF** [comment]
> Please rename dependencies to versions.

Okay.

> Im not a fan of including any dependency in our responses. This would be another operatioon of the generator API and the xcube server API.

I changed this, too, but to me this directly contradicts the statement in the issue:

> Should be included in responses of
xcube serve
xcube gen2

File: xcube/core/gen2/response.py [hunk 9]
```diff
@@ -52,6 +53,7 @@ def __init__(self,
         self.message = message if message else None
         self.output = list(output) if output else None
         self.traceback = list(traceback) if traceback else None
+        self.versions = versions if versions else None
```
Comment (forman) [inline xcube/core/gen2/response.py:56]:
```suggestion
        self.versions = dict(versions) if versions else None
```

File: xcube/util/versions.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
+# The MIT License (MIT)
+# Copyright (c) 2019 by the xcube development team and contributors
```
Comment (forman) [inline xcube/util/versions.py:2]:
```suggestion
# Copyright (c) 2021 by the xcube development team and contributors
```

File: test/util/test_dependencies.py [hunk 11]
```diff
@@ -0,0 +1,19 @@
+import unittest
```
Comment (forman) [inline test/util/test_dependencies.py:1]: Please rename `test_dependencies.py` into `test_versions.py`

File: xcube/webapi/service.py [hunk 12]
```diff
@@ -303,6 +304,7 @@ def write_error(self, status_code, **kwargs):
                 'error': {
                     'code': status_code,
                     'message': self._reason,
+                    'versions': get_xcube_versions(),
```
Comment (forman) [inline xcube/webapi/service.py:307]: Maybe not recompute on every error.

File: xcube/core/gen2/remote/generator.py [hunk 13]
```diff
@@ -316,6 +317,7 @@ def _new_invalid_result(cls, state: CubeGeneratorState):
             status_code=422,
             message='missing cube generator result',
             output=state.output,
+            versions=get_xcube_versions()
```
Comment (forman) [inline xcube/core/gen2/remote/generator.py:320]: Maybe not recompute on every error.

**Review by forman — CHANGES_REQUESTED** [review]
Sorry, still cannot approve it, some minor issues remain.

Maybe we should not recompute the versions dicts on every error, because 
1. it may cost time to load modules that haven't been loaded yet
2. it is not impossible that `get_xcube_version()` raises while we construct an error result for another error

**Review by forman — APPROVED** [review]

**Pull request merged**